File: hackerrank/problem solving/algorithms/2 - implementation/53 - flatland space stations/flatland_space_stations_procedural.java
// Source: https://www.hackerrank.com/challenges/flatland-space-stations/problem?isFullScreen=true

import java.util.Scanner;
import java.util.List;
import java.util.Arrays;
import java.util.stream.Collectors;
import java.util.Collections;
import java.lang.Math;

public class FlatlandSpaceStations {
    private static Scanner _scanner;

    public static void main(String[] args) {
        _scanner = new Scanner(System.in);

        final int nCities = _scanner.nextInt(), nCitiesWithSpaceStation = _scanner.nextInt();
        final List<Integer> citiesWithSpaceStation = _readAnIntArray(nCitiesWithSpaceStation);
        System.out.println(_findMaxDistanceFromSpaceStation(nCities, citiesWithSpaceStation));
    }

        private static List<Integer> _readAnIntArray(int size) {
            return Arrays.asList(new Integer[size]).stream().map(element -> _scanner.nextInt()).collect(Collectors.toList());
        }

        private static int _findMaxDistanceFromSpaceStation(int nCities, List<Integer> citiesWithSpaceStation) {
            Collections.sort(citiesWithSpaceStation);
            final int firstCity = 0;
            int maxDistance = citiesWithSpaceStation.get(0) - firstCity;
            final int size = citiesWithSpaceStation.size();

            for (int i = 1, previousCity = citiesWithSpaceStation.get(0); i < size; i++) {
                final int distance = _calculateDistanceBetweenCities(citiesWithSpaceStation.get(i), previousCity);
                maxDistance = Math.max(maxDistance, distance);
                previousCity = citiesWithSpaceStation.get(i);
            }

            final int lastCity = nCities - 1;
            final int lastCityWithSpaceStation = citiesWithSpaceStation.get(size - 1);
            if (!_hasLastCitySpaceStation(lastCity, lastCityWithSpaceStation)) {
                final int distance = _calculateDistanceOfLastCity(lastCity, lastCityWithSpaceStation);
                maxDistance = Math.max(maxDistance, distance);
            }

            return maxDistance;
        }

            private static int _calculateDistanceBetweenCities(int cityWithSpaceStation, int previousCity) {
                return (cityWithSpaceStation - previousCity) / 2;
            }

            private static boolean _hasLastCitySpaceStation(int lastCity, int lastCityWithSpaceStation) {
                return lastCity == lastCityWithSpaceStation;
            }

            private static int _calculateDistanceOfLastCity(int lastCity, int lastCityWithSpaceStation) {
                return lastCity - lastCityWithSpaceStation;
            }
}

File: hackerrank/problem solving/algorithms/2 - implementation/53 - flatland space stations/flatland_space_stations_procedural.cpp
// Source: https://www.hackerrank.com/challenges/flatland-space-stations/problem?isFullScreen=true

#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

int find_max_distance_from_space_station(const int n_cities, vector<int>& cities_with_space_station);
    int calculate_distance_between_cities(const int city_with_space_station, const int previous_city);
    bool has_last_city_space_station(const int last_city_with_space_station, const int last_city);
    int calculate_distance_of_last_city(const int last_city_with_space_station, const int last_city);

int main()
{
    int n_cities, n_cities_with_space_station;
    cin >> n_cities >> n_cities_with_space_station;

    vector<int> cities_with_space_station(n_cities_with_space_station);
    ranges::generate(cities_with_space_station, [] {int element; cin >> element; return element;});

    cout << find_max_distance_from_space_station(n_cities, cities_with_space_station) << "\n";

    return 0;
}

    int find_max_distance_from_space_station(const int n_cities, vector<int>& cities_with_space_station)
    {
        ranges::sort(cities_with_space_station);
        const int first_city {0};
        int max_distance {cities_with_space_station.front() - first_city};

        for (int i {1}, previous_city {cities_with_space_station.front()}, size {static_cast<int>(cities_with_space_station.size())}; i < size; ++i) {
            const int distance {calculate_distance_between_cities(cities_with_space_station.at(i), previous_city)};
            max_distance = {max(max_distance, distance)};
            previous_city = {cities_with_space_station.at(i)};
        }

        if (int last_city {n_cities - 1}, last_city_with_space_station {cities_with_space_station.back()}; !has_last_city_space_station(last_city, last_city_with_space_station)) {
            const int distance {calculate_distance_of_last_city(last_city, last_city_with_space_station)};
            max_distance = {max(max_distance, distance)};
        }

        return max_distance;
    }

        int calculate_distance_between_cities(const int city_with_space_station, const int previous_city)
        {
            return (city_with_space_station - previous_city) / 2;
        }

        bool has_last_city_space_station(const int last_city, const int last_city_with_space_station)
        {
            return last_city_with_space_station == last_city;
        }

        int calculate_distance_of_last_city(const int last_city, const int last_city_with_space_station)
        {
            return last_city - last_city_with_space_station;
        }
